### PR TITLE
[Review Only] Add WireTap Support

### PR DESCRIPTION
**Pull request by garyrussell**
Notice the one test that doesn't work.

I am not sure it's worth bending over backwards to support that one case; we would have to somehow defer attaching the interceptor.

### CODE REVIEW

File: src/main/java/org/springframework/integration/dsl/channel/DslWireTap.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+import org.springframework.messaging.core.DestinationResolutionException;
+import org.springframework.messaging.support.ChannelInterceptor;
+import org.springframework.util.Assert;
+
+/**
+ * @author Gary Russell
+ * @since 1.0.2
+ *
+ */
+public class DslWireTap implements ChannelInterceptor, BeanFactoryAware, VetoCapableInterceptor, InitializingBean,
```
Comment (artembilan) [inline src/main/java/org/springframework/integration/dsl/channel/DslWireTap.java]: I make similar DSL-specific classes as `package protected` do not confuse end-users

File: src/main/java/org/springframework/integration/dsl/IntegrationFlowDefinition.java [hunk 2]
```diff
@@ -231,6 +234,27 @@ public B publishSubscribeChannel(Executor executor,
 		return addComponents(spec.getComponentsToRegister()).channel(spec);
 	}
 
+	public B tap(String tapChannel) {
```
Comment (artembilan) [inline src/main/java/org/springframework/integration/dsl/IntegrationFlowDefinition.java]: I don't think that this is a good name. `wireTap` is wire-tap in terms of SI.

File: src/main/java/org/springframework/integration/dsl/channel/WireTapSpec.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+import org.springframework.integration.filter.ExpressionEvaluatingSelector;
+import org.springframework.messaging.MessageChannel;
+import org.springframework.util.Assert;
+
+/**
+ * @author Gary Russell
+ * @since 1.0.2
+ *
+ */
+public class WireTapSpec extends IntegrationComponentSpec<WireTapSpec, DslWireTap> implements ComponentsRegistration {
```
Comment (artembilan) [inline src/main/java/org/springframework/integration/dsl/channel/WireTapSpec.java]: No one `Spec` should provide `public` ctor

Comment (artembilan) [inline src/main/java/org/springframework/integration/dsl/channel/WireTapSpec.java]: No one `Spec` should expose internal classes like this `DslWireTap`.
See `RecipientListRouterSpec`

**Comment by artembilan** [comment]
This s a good request, Gary. Thanks.
I didn't think before that it would make sence, but yes, any anonymous channels between endpoint should have `wire-tap` option as well. And this one EIP-method does the stuff.
However now I see a bit different solution without that `DslWireTap`, plus as `sub-flow` variant.
I understand the issue with just channel name and I think we can overcome that with an internal sub-flow.
If you want I can all my thoughts as polishing to your PR (or as a new one).
Otherwise let's go with a long discussion during your polishing according to my comments.

**Comment by artembilan** [comment]
Here is what I think on the matter https://github.com/artembilan/spring-integration-java-dsl/commit/75525f3dcd730e34adf8b6b8265258a1de8121ec
Let me know if you are OK and I go ahead with JavaDocs and Docs.

Thank you for pointing that out anyway!

**Comment by garyrussell** [comment]
Looks good! Thanks.

**Comment by artembilan** [comment]
The JavaDocs for your confirmation https://github.com/artembilan/spring-integration-java-dsl/commit/d3bbd8aa0bfa51bd2eddfdb243a395f2e6196f28

**Comment by artembilan** [comment]
@garyrussell , would be great, if you take a look to my last JavaDocs polishing.
Thanks

**Comment by garyrussell** [comment]
LGTM; please go ahead and merge.

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-integration-java-dsl/commit/06e9c477e57c37571d2d2857272e8016746b0a09

**Pull request closed without merging**